Final functionality.

alexanderstrelkov opened the pull request.

alexanderstrelkov committed:
Deleted spaces, cloned project

alexanderstrelkov committed:
Merge branch 'feature/-cloning-project' into develop

alexanderstrelkov committed:
Made label, button, textField and passGenerator.

alexanderstrelkov committed:
Merge branch 'feature/creating-UI-Elemens' into develop

alexanderstrelkov committed:
Made generatePass, searchPass and clear button with logic.

alexanderstrelkov committed:
Merge branch 'feature/bruteForce-logic' into develop

alexanderstrelkov committed:
Made stop bruteForce method.

alexanderstrelkov committed:
Merge branch 'feature/extra-task-logic' into develop

Artmukho commented on iOS6-HW15-Alexander.Strelkov/ViewController.swift:
```diff
@@ -8,12 +8,122 @@
 class ViewController: UIViewController {
-
+    
+    var isStopped: Bool = false
+        
+    @IBOutlet weak var textField: UITextField!
+    @IBOutlet weak var activityIndicator: UIActivityIndicatorView!
+    @IBOutlet weak var label: UILabel!
+    @IBAction func searchPasswordButton(_ sender: UIButton) {
+        self.stopButton.isEnabled = true
+        self.bruteForce(passwordToUnlock: textField.text ?? "")
+    }
+    @IBAction func generateButton(_ sender: UIButton) {
+        textField.text = "\(randomPasswordGenerate())"
+    }
+    @IBOutlet weak var clearButton: UIButton!
+    @IBAction func clearButton(_ sender: UIButton) {
+        textField.text = ""
+        label.text = ""
+        textField.isSecureTextEntry = true
+        isStopped = false
+        clearButton.isEnabled = false
+    }
+    @IBOutlet weak var stopButton: UIButton!
+    @IBAction func stopButton(_ sender: UIButton) {
+        isStopped = true
+    }
+    
```
Мешанина какая-то, будет здорово это структурировать. Чтобы аутлеты с аутлетами, экшены с экшенами.

Artmukho commented on iOS6-HW15-Alexander.Strelkov/ViewController.swift:
```diff
@@ -8,12 +8,122 @@
+        stopButton.isEnabled = false
+    }
+    
+    func randomPasswordGenerate() -> String {
+        let passwordLength = 3
+        let passwordCharacters =
+        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
+        let randomPassword = String((0..<passwordLength).compactMap{ _ in passwordCharacters.randomElement() })
+        return randomPassword
+    }
```
Отличная функция. Пароль рандомится))

Artmukho commented on iOS6-HW15-Alexander.Strelkov/ViewController.swift:
```diff
@@ -8,12 +8,122 @@
     }
+}
+
+extension String {
+    var digits: String { return "0123456789" }
+    var lowercase: String { return "abcdefghijklmnopqrstuvwxyz" }
+    var uppercase: String { return "ABCDEFGHIJKLMNOPQRSTUVWXYZ" }
+    var punctuation: String { return "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~" }
+    var letters: String { return lowercase + uppercase }
+    var printable: String { return digits + letters + punctuation }
+    
+    mutating func replace(at index: Int, with character: Character) {
+        var stringArray = Array(self)
+        stringArray[index] = character
+        self = String(stringArray)
+    }
+}
```
Этому расширению место в другом файле)

Artmukho commented on iOS6-HW15-Alexander.Strelkov/ViewController.swift:
```diff
@@ -8,12 +8,122 @@
+}
 
+func indexOf(character: Character, _ array: [String]) -> Int {
+    return array.firstIndex(of: String(character))!
```
Принудительное извлечение опционала. О порядке в проекте, я говорил в воскресенье. Также говорил на что обратить внимание и чего быть не должно.

Artmukho commented on iOS6-HW15-Alexander.Strelkov/ViewController.swift:
```diff
@@ -8,12 +8,122 @@
+    }
+    else {
+        str.replace(at: str.count - 1,
+                    with: characterAt(index: (indexOf(character: str.last!, array) + 1) % array.count, array))
```
Форс

Artmukho commented on iOS6-HW15-Alexander.Strelkov/ViewController.swift:
```diff
@@ -8,12 +8,122 @@
+        str.replace(at: str.count - 1,
+                    with: characterAt(index: (indexOf(character: str.last!, array) + 1) % array.count, array))
+        
+        if indexOf(character: str.last!, array) == 0 {
+            str = String(generateBruteForce(String(str.dropLast()), fromArray: array)) + String(str.last!)
```
2 форса))

Artmukho commented on iOS6-HW15-Alexander.Strelkov/ViewController.swift:
```diff
@@ -8,12 +8,122 @@
+    }
+}
 
+func indexOf(character: Character, _ array: [String]) -> Int {
+    return array.firstIndex(of: String(character))!
+}
 
+func characterAt(index: Int, _ array: [String]) -> Character {
+    return index < array.count ? Character(array[index])
+    : Character("")
 }
 
+func generateBruteForce(_ string: String, fromArray array: [String]) -> String {
+    var str: String = string
+    
+    if str.count <= 0 {
+        str.append(characterAt(index: 0, array))
+    }
```
Не есть хорошо делать эти функции в глобальном доступе

Artmukho commented on iOS6-HW15-Alexander.Strelkov/ViewController.swift:
```diff
@@ -8,12 +8,122 @@
+    
+    func bruteForce(passwordToUnlock: String) {
+        let mainQueue = DispatchQueue.main
+        mainQueue.async {
+            self.activityIndicator.isHidden = false
+            self.activityIndicator.startAnimating()
+        }
```
Этот код и так выполнится на главном потоке

Artmukho commented on iOS6-HW15-Alexander.Strelkov/ViewController.swift:
```diff
@@ -8,12 +8,122 @@
+            var password: String = ""
+            while password != passwordToUnlock && !self.isStopped {
+                password = generateBruteForce(password, fromArray: ALLOWED_CHARACTERS)
+                mainQueue.async {
+                    self.label.text = "Hacking.. \(password)"
+                    if password == passwordToUnlock {
+                        self.activityIndicator.stopAnimating()
+                        self.activityIndicator.isHidden = true
+                        self.textField.isSecureTextEntry = false
+                        self.label.text = "Hacked! Password: \(password)"
+                        self.clearButton.isEnabled = true
+                        self.stopButton.isEnabled = false
+                    }
+                }
+                if self.isStopped {
+                    mainQueue.async {
+                        self.activityIndicator.isHidden = true
+                        self.label.text = "Password \(password) is not hacked!"
+                        self.clearButton.isEnabled = true
+                    }
```
А вот тут вызов главного потока в тему, так как нам нужно обновлять UI

Artmukho reviewed (commented):
Привет, Сашь! Проверил твою работу. Основное задание выполнено. Замечания по коду оставил в комментариях, а также:

1. Нет структуры кода. Распредели файлы по папкам. Также вынеси некоторый код из контроллера. Писать все в одном файле не есть хорошо.
2.  Стоит добавить защиту от дураков. Если нажать на кнопку search password до генерации. Это все ломает немного логику. По нажатию на стоп анимация индикатора продолжается

Жду твою работу с исправлениями! Успехов в выполнении задания!

alexanderstrelkov committed:
Optimized code

removed methods from VC to separate file, remade searchPass func logic if there is no pass to hack

alexanderstrelkov committed:
Merge branch 'buf-fix/optimising-code' into develop

alexanderstrelkov committed:
Bugs Fixed

alexanderstrelkov committed:
Merge branch 'buf-fix/optimising-code' into develop
